### PR TITLE
Add new lint: `used_underscore_items`

### LINKED ISSUE #13260: Lint use of functions with underscore-prefixed names
### What it does

Flag use of functions with underscore-prefixed names.  See https://github.com/rust-lang/rust-clippy/discussions/13256

### Advantage

Same as the [existing `used_underscore_binding` lint](https://rust-lang.github.io/rust-clippy/stable/index.html#/used_underscore_binding).

### Drawbacks

_No response_

### Example

```rust
fn _foo() {
  println!("Foo");
}

fn main() {
  _foo();
}
```

Could be written as:

```rust
fn foo_() {
  println!("Foo");
}

fn main() {
  foo_();
}
```

### PR DESCRIPTION
**Pull request by WeiTheShinobi**
Closes #13260 

---

changelog: new [`used_underscore_items`] lint against using items with a single leading underscore

### CODE REVIEW

File: clippy_lints/src/misc.rs [hunk 1]
```diff
@@ -205,51 +245,105 @@ impl<'tcx> LateLintPass<'tcx> for LintPass {
...
+        && !fulfill_or_allowed(cx, USED_UNDERSCORE_BINDING, [expr.hir_id, definition_hir_id])
+    {
+        span_lint_and_then(
+            cx,
+            USED_UNDERSCORE_BINDING,
+            expr.span,
+            format!(
+                "used binding `{name}` which is prefixed with an underscore. A leading \
+                 underscore signals that a binding will not be used"
+            ),
```
Comment (llogiq) [inline clippy_lints/src/misc.rs:342]: That message is quite long for something so insignificant. Let's remove the second part, people can read it in the docs if needed.

**Comment by y21** [comment]
FWIW, pretty much almost all of the x86 platform intrinsics are prefixed with an underscore: https://doc.rust-lang.org/stable/core/arch/x86_64/index.html

Seems unfortunate if we start linting on any use of them... :thinking:

**Comment by llogiq** [comment]
I think a good balance here would be to check only the use of intra-crate items (that is items from the current crate), as the user cannot control how others name their items.

**Comment by llogiq** [comment]
Thank you!

@bors r+

**Pull request merged**